### DISCUSSION TITLE
DTS include dependencies

### DISCUSSION THREAD

**Discussion by Dennis1022** [mesonbuild/meson#9724] [Q&A] [answered]
As part of my Meson build I am generating a DTB from a DTS which can of course include other .dtsi files. Because the device tree compiler does not natively support #include, I have a meson custom target where I run the c pre processor before passing this file to the the device tree compiler. My pre processing step has a dependency on the root DTS file, but knows nothing about the included .dtsi files. Does any one have a suggestion on how to resolve this? Currently if one of the included .dtsi files changes, it does not trigger a rebuild.

My initial thought was to instruct the c pre processor to generate a .d file where all the dependencies would be listed. With this I could use the fs module to read and create a file list. Unfortunately, I am unable to use the fs read() method in the build tree:

`meson.build:53:0: ERROR: path must not be in the build tree`

**Comment by eli-schwartz** [comment, marked as answer]
Why do you need to generate a file list? The fs module runs during the configure stage, before the .d file is created, which is why it is not permitted to read files in the build tree. But, the point of .d file dependencies is for *dynamic* detection of rebuild dependencies, so the intended way to handle this would be for the .d file to be emitted to the `depfile` kwarg of e.g. custom_target.

Meson would then record this in the build.ninja as a dynamic runtime dependency file list, and ninja would absorb and then delete the depfile as soon as it is created.

**Comment by Dennis1022** [comment]
Sounds like I don't need to 😄. I was originally thinking I needed to in order to use the `depend_files` kwargs of custom_target. Thank you for pointing out the `depfile` kwarg, this gives me the behavior I was looking for!
